src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -133,38 +133,126 @@ private void EmbedIfNeedTo(BoundExpression receiver, ImmutableArray<PropertySymb
+                return;
+            }
+
+            // PROTOTYPE(ic): should we also reduce both methods if 'method' is being called as an extension method?
+
+            method.TryGetThisParameter(out var methodThisParameter);
+            symbolForCompare.TryGetThisParameter(out var interceptorThisParameterForCompare);
+            switch (methodThisParameter, interceptorThisParameterForCompare)
+            {
+                case (not null, null):
+                case (not null, not null) when !methodThisParameter.Type.Equals(interceptorThisParameterForCompare.Type, TypeCompareKind.ObliviousNullableModifierMatchesAny)
+                        || methodThisParameter.RefKind != interceptorThisParameterForCompare.RefKind: // PROTOTYPE(ic): and ref custom modifiers are equal?
+                    this._diagnostics.Add(ErrorCode.ERR_InterceptorMustHaveMatchingThisParameter, interceptsLocationAttributeData.AttributeLocation, methodThisParameter, method);
+                    break;
+                case (null, not null):
+                    this._diagnostics.Add(ErrorCode.ERR_InterceptorMustNotHaveThisParameter, interceptsLocationAttributeData.AttributeLocation, method);
+                    break;
+                default:
+                    break;
+
+            }
+
+            if (needToReduce)
+            {
+                Debug.Assert(methodThisParameter is not null);
+                arguments = arguments.Insert(0, receiverOpt!);
+                receiverOpt = null;
+
+                // CodeGenerator.EmitArguments requires that we have a fully-filled-out argumentRefKindsOpt for any ref/in/out arguments.
+                var thisRefKind = methodThisParameter.RefKind;
+                if (argumentRefKindsOpt.IsDefault && thisRefKind != RefKind.None)
+                {
+                    argumentRefKindsOpt = method.Parameters.SelectAsArray(static param => param.RefKind);
+                }
+
+                if (!argumentRefKindsOpt.IsDefault)
+                {
+                    argumentRefKindsOpt = argumentRefKindsOpt.Insert(0, thisRefKind);
+                }
+
+                // PROTOTYPE(ic): search for a scenario where propagating this value matters
+                invokedAsExtensionMethod = true;
+            }
+
+            method = interceptor;
+
+            return;
+        }
+
         public override BoundNode VisitCall(BoundCall node)
         {
             Debug.Assert(node != null);
 
+            // PROTOTYPE(ic): interception should occur after lowering
+            // so that we use the original method in all relevant situations such as interpolated string lowering.
+            var (method, receiverOpt, arguments, argsToParamsOpt, argRefKindsOpt, invokedAsExtensionMethod) = (node.Method, node.ReceiverOpt, node.Arguments, node.ArgsToParamsOpt, node.ArgumentRefKindsOpt, node.InvokedAsExtensionMethod);
```

Comment:
>var (method, receiverOpt, arguments, argsToParamsOpt, argRefKindsOpt, invokedAsExtensionMethod) = (node.Method, node.ReceiverOpt, node.Arguments, node.ArgsToParamsOpt, node.ArgumentRefKindsOpt, node.InvokedAsExtensionMethod); [](http://example.com/codeflow?start=12&length=225)

This line is not readable. And also doesn't follow the `var` guidance. Consider declaring each local separately with explicit type and initializer.